### PR TITLE
Make sure the URL installer does not allow other schemas then http and https

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for an Issue raised to the JSST.

### Summary of Changes

Make sure the URL installer does not allow other schemas then http and https

### Testing Instructions

- Try to install an extension form this url: `ftp://joomla.zip`
- Joomla tires to contact an FTP server
- apply this patch
- try `ftp://joomla.zip` again.
- There is now a dedicated message and we dont try to contact an FTP server
- make sure that extension installation from URL still works as before.

### Actual result BEFORE applying this Pull Request

There is an message but we still try to contact the FTP server

### Expected result AFTER applying this Pull Request

There is now a dedicated message and we dont try to contact an FTP server

### Documentation Changes Required

none

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 1]
```diff
@@ -123,6 +123,7 @@ COM_INSTALLER_MSG_DISCOVER_PURGEDDISCOVEREDEXTENSIONS="Cleared discovered extens
 COM_INSTALLER_MSG_ERROR_CANT_CONNECT_TO_UPDATESERVER="Can't connect to %s"
 COM_INSTALLER_MSG_INSTALL_ENTER_A_URL="Please enter a URL"
 COM_INSTALLER_MSG_INSTALL_INVALID_URL="Invalid URL"
+COM_INSTALLER_MSG_INSTALL_INVALID_URL_SCHEMA="Please enter a valid URL starting with http or https"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]:
```suggestion
COM_INSTALLER_MSG_INSTALL_INVALID_URL_SCHEMA="Please enter a valid URL starting with http or https."
```
Its a sentence so it needs to end with a .

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on eae250f4040b68825a2cc1982915a4e1ddfb6342<br /><br />Tested successfully in 3.9.25-dev of 20 January using PHP 8.0.1.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32087">issues.joomla.org/tracker/joomla-cms/32087</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on eae250f4040b68825a2cc1982915a4e1ddfb6342<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32087">issues.joomla.org/tracker/joomla-cms/32087</a>.</sub>

**Comment by PhilETaylor** [comment]
Lol when I reported this years ago I was told it was a non-problem and only idiots would try to use a non http prefix ...

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 2]
```diff
@@ -123,6 +123,7 @@ COM_INSTALLER_MSG_DISCOVER_PURGEDDISCOVEREDEXTENSIONS="Cleared discovered extens
 COM_INSTALLER_MSG_ERROR_CANT_CONNECT_TO_UPDATESERVER="Can't connect to %s"
 COM_INSTALLER_MSG_INSTALL_ENTER_A_URL="Please enter a URL"
 COM_INSTALLER_MSG_INSTALL_INVALID_URL="Invalid URL"
+COM_INSTALLER_MSG_INSTALL_INVALID_URL_SCHEMA="Please enter a valid URL starting with http or https."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_installer.ini]: `SCHEMA` or `SCHEME`?

Comment (PhilETaylor) [reply administrator/language/en-GB/en-GB.com_installer.ini]: "SCHEME" is the correct English word

http is a SCHEME 

https://en.wikipedia.org/wiki/List_of_URI_schemes

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_installer.ini]: schema is what structured data is - schemas.org

Comment (PhilETaylor) [reply administrator/language/en-GB/en-GB.com_installer.ini]: The validation rule is not testing the "schema" (the structured data) of a domain name, its testing the `scheme` part of the URL, just the scheme, and therefore the validation error message should refer to the SCHEME and not the schema. 

It even states that in the php function 

```php
if (!in_array($uri->getScheme(), ['http', 'https'])
```

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_installer.ini]: I'm agreeing with you

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_installer.ini]:
```suggestion
COM_INSTALLER_MSG_INSTALL_INVALID_URL_SCHEME="Please enter a valid URL starting with http or https."
```

File: administrator/components/com_installer/models/install.php [hunk 3]
```diff
@@ -385,6 +385,16 @@ protected function _getPackageFromUrl()
 			return false;
 		}
 
+		// We only allow http & https here
+		$uri = new JUri($url);
+
+		if (!in_array($uri->getScheme(), ['http', 'https']))
+		{
+			JError::raiseWarning('', JText::_('COM_INSTALLER_MSG_INSTALL_INVALID_URL_SCHEMA'));
```
Comment (zero-24) [inline administrator/components/com_installer/models/install.php]:
```suggestion
			JError::raiseWarning('', JText::_('COM_INSTALLER_MSG_INSTALL_INVALID_URL_SCHEME'));
```

**Comment by zero-24** [comment]
> Lol when I reported this years ago I was told it was a non-problem and only idiots would try to use a non http prefix ...

Well its still not handled as security issue but it was reported as one. ;) Given that we had some kind of not working JS "validation" we choose to move this forward to the public tracker and get it fixed anyway.

File: administrator/language/en-GB/en-GB.com_installer.ini (hunk 2 above)
Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_installer.ini]: Thanks patched

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on e393a65ca6e84ef8ba250bd28ecc3b60a4502dee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32087">issues.joomla.org/tracker/joomla-cms/32087</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on e393a65ca6e84ef8ba250bd28ecc3b60a4502dee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32087">issues.joomla.org/tracker/joomla-cms/32087</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32087">issues.joomla.org/tracker/joomla-cms/32087</a>.</sub>

**Comment by drmenzelit** [comment]
Thanks

**Comment by zero-24** [comment]
Thanks @drmenzelit

**Pull request merged**